irksome/nprk.py

```diff
@@ -0,0 +1,328 @@
+                                 implicit solves in the first argument.
+    :arg arg2_solver_parameters: A :class:`dict` of solver parameters that will be used in the
+                                 implicit solves in the second argument.
+    :arg mass_parameters: A :class:`dict` of solver parameters that will be used in the mass
+                          solves when the operator must be evaluated on previous stages.
+    :arg nullspace: An optional null space object.
+    """
+
+    def __init__(self, F, nprk_tableau: NPRKTableau, t, dt, u0, u1, u2, bcs=None,
+                 arg1_solver_parameters=None, arg2_solver_parameters=None,
+                 mass_parameters=None, appctx=None, nullspace=None):
+
+        self.num_steps = 0
+
+        # Number of non-linear and linear iterations in solves in each argument
+        self.num_nonlinear_iterations = np.array([0, 0], dtype=int)
+        self.num_linear_iterations = np.array([0, 0], dtype=int)
+
+        self.num_mass_nonlinear_iterations = 0
+        self.num_mass_linear_iterations = 0
+
+        self.tableau = nprk_tableau
+
+        self.dt = dt
+        self.u = u0
+
+        # (NOTE): Can use list of time-dependent variables to reduce the length of the loops
+        #         when summing over previous stages for a potential performance improvement
+        _ = _check_aux_t_dep(F, u0, u1, u2)
+
+        self.u_stage, self.u_tilde, self.u_exp, self.ac, F_arg1, F_arg2, \
+            self.k, self.u_arg1, self.u_arg2, F_mass = getFormsNPRK(F, u0, u1, u2, t, dt)
+
+        V = u0.function_space()
+        # Operator evaluations for the pairs of stage arguments
+        self.fs = {tuple(jk): Function(V) for jk in nprk_tableau.all_indices}
+        # Stage values, needed for most general tableaux
+        self.us = [Function(V) for _ in range(nprk_tableau.num_stages)]
+
+        # BCs can be passed straight through to stage value solvers
+        arg1_prob = NonlinearVariationalProblem(F_arg1, self.u_stage, bcs)
+        arg2_prob = NonlinearVariationalProblem(F_arg2, self.u_stage, bcs)
+
+        # Additional BCs for mass solvers
+        bcs_mass = []
+        for bc in bcs:
+            bcs_mass.append(bc.reconstruct(g=0.0))
+
+        mass_prob = NonlinearVariationalProblem(F_mass, self.k, bcs_mass)
+
+        appctx_irksome = {"F": F,
+                          "nprk_tableau": nprk_tableau,
+                          "t": t,
+                          "dt": dt,
+                          "u0": u0,
+                          "bcs": bcs,
+                          "bc_type": "DAE",
+                          "stage_type": "nprk",
+                          "nullspace": nullspace}
+        if appctx is None:
+            appctx = appctx_irksome
+        else:
+            appctx = {**appctx, **appctx_irksome}
+
+        # One solver corresponding to each implicit argument
+        self.solvers = {}
```

Comment:
Parallel construction -- you pass two sets of parameters with names, but then you stash the solvers in a dict.  Also, why use a dict if you're just indexing it with 0, 1?  A list/tuple would be fine.

And if you had the args come in via an iterable, creating the problems/solvers would also be done with a comprehension..